dhcpv4/nclient4/client.go

```diff
@@ -627,6 +609,35 @@ func (c *Client) SendAndRead(ctx context.Context, dest *net.UDPAddr, p *dhcpv4.D
 	return response, nil
 }
 
+// TransactAckNak exposes a higher-level API than SendAndRead for sending a given request (any type)
+// and receiving an ACK/NAK response. The server identifier is validated (if given), responses are
+// type checked, and NAK responses are automatically returned as errors.
+func (c *Client) TransactAckNak(ctx context.Context, request *dhcpv4.DHCPv4) (*dhcpv4.DHCPv4, error) {
+	matchers := []Matcher{IsMessageType(dhcpv4.MessageTypeAck, dhcpv4.MessageTypeNak)}
+
+	// Servers are supposed to only respond to Requests containing their server identifier,
+	// but sometimes non-compliant servers respond anyway.
+	// Clients are not required to validate this field. Servers are required to include
+	// the server identifier in their Offer per RFC 2131 Section 4.3.1 Table 3, but it is
+	// not mandatory for e.g. INFORM requests if the client does not explicitly request it.
+	identifier := request.ServerIdentifier()
+	if identifier != nil {
```

Comment:
This logic is changing from what we had in `RequestFromOffer`. Is there any reason for it? When will a Servicer Indentifier be nil?

Reply:
This change is to facilitate compatbility with INFORM requests. The comment right above the line to it explains the reason for why `ServerIdentifier` can be `nil`:

> Servers are required to include the server identifier in their Offer per RFC 2131 Section 4.3.1 Table 3, but it is not mandatory for e.g. INFORM requests if the client does not explicitly request it.

Reply:
If the server has chosen not to send the identifier, it cannot be validated.

Reply:
This is changing the previous behavior and relaxing the checks. If `RequestFromOffer` does not get a `ServerIdentifier` back, with this change it will not raise an error.

> If the server has chosen not to send the identifier

I thought this was not optional. Do you have any reference to the RFC about this?

Reply:
You're right, it seems to not be optional. The server must always send the identifier for ACK/NAK as per RFC 2131 Section 4.3.1 Table 3, and DHCPINFORM follows the ACK parameters as defined in [Section 4.3.5](https://www.rfc-editor.org/rfc/rfc2131#section-4.3.5). Will fix.